src/duckling/canvas/map-editor.component.ts

```diff
@@ -199,98 +172,146 @@ export class MapEditorComponent implements AfterViewInit, OnDestroy {
 
     onToolSelected(newTool : BaseTool) {
         this._setTool(newTool);
-        this._redrawAllDisplayObjects();
+        this._markCacheDirty(TOOL_CACHE_KEY)
     }
 
     onStageDimensonsChanged(stageDimensions : Vector) {
         this.projectService.changeDimension(stageDimensions);
-        this._redrawAllDisplayObjects();
+        this._markCacheDirty(BACKGROUND_CACHE_KEY);
+        this._markCacheDirty(GRID_CACHE_KEY);
     }
 
     onGridSizeChanged(gridSize : number) {
         this.projectService.changeGridSize(gridSize);
-        this._gridDisplayObject = this._buildGrid();
-        this.canvasDisplayObject = this._buildCanvasDisplayObject();
+        this._markCacheDirty(GRID_CACHE_KEY);
     }
 
     onScaleChanged(scale : number) {
         this.scale = scale;
-        this._redrawAllDisplayObjects();
+        this._markAllCachesDirty();
     }
 
     onShowGridChanged(showGrid : boolean) {
         this.showGrid = showGrid;
-        this.canvasDisplayObject = this._buildCanvasDisplayObject();
+        this._markCacheDirty(GRID_CACHE_KEY);
     }
 
-    openMap(mapKey : string) {
-        this.projectService.openMap(mapKey);
+    async openMap(mapKey : string) {
+        await this.projectService.openMap(mapKey);
+        this._markAllCachesDirty();
     }
 
     private _redrawAllDisplayObjects() {
-        this._canvasBackgroundDisplayObject = this._buildCanvasBackground();
-        this._canvasBorderDisplayObject = this._buildCanvasBorder();
-        this._gridDisplayObject = this._buildGrid();
-        this._toolDisplayObject = this.tool.getDisplayObject(this.scale);
+        if (this._drawableCache[BACKGROUND_CACHE_KEY].dirty) {
```

Comment:
I think DrawableCache should be a class. The constructor should take a function that is called to rebuild the cache. You ould transform things like:

this._markeCacheDirty(cacheKey) to this.cache.cacheKey.dirty = true;